### PR TITLE
Model decorator

### PR DESCRIPTION
**Pull request by sdelamo**
What do you (@graemerocher @jameskleeh ) think?

### CODE REVIEW

**Comment by jameskleeh** [comment]
`Map<String, Object> modelForRequest(HttpRequest request);`

This would limit to adding to the model. I would make it so parts of the model can be removed etc.

`void decorateModel(HttpRequest request, Map<String, Object> model)`

**Comment by sdelamo** [comment]
I've modify the interface to use @jameskleeh suggestion.

File: views/src/main/java/io/micronaut/views/ViewsFilter.java [hunk 1]
```diff
@@ -48,21 +52,25 @@
...
+    private final Collection<ViewsModelDecorator> modelDecorators;
 
     /**
      * Constructor.
      *
      * @param beanLocator The bean locator
      * @param viewsFilterOrderProvider The order provider
+     * @param modelDecorators Collection of views model decorator beans
      */
     public ViewsFilter(BeanLocator beanLocator,
```
Comment (jameskleeh) [inline views/src/main/java/io/micronaut/views/ViewsFilter.java]: You will need to deprecate this constructor and create a new one

**Comment by graemerocher** [comment]
I made some modifications:

* Rename `ViewModelDecorator` to `ViewModelProcessor` and change signature
* Now accepts a `ModelAndView` so implementors can potentially alter the view to be rendered (more flexibility)

Thoughts?

File: views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * @author Sergio del Amo
+ * @since 1.0
+ */
+public interface ViewsModelDecorator {
+
+    /**
+     * @param model Model being processed
+     * @param request The request being processed
+     */
+    void decorateModel(@Nonnull Map<String, Object> model, @Nonnull HttpRequest request);
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java]: @graemerocher has renamed it to `process`

File: views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+/**
+ * Implementers of {@link ViewsModelDecorator} decorate response model with additional entries.
+ *
+ * @author Sergio del Amo
+ * @since 1.0
+ */
+public interface ViewsModelDecorator {
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java]: @graemerocher renamed to `ViewModelProcessor`

File: views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+package io.micronaut.views.model;
+
+import io.micronaut.http.HttpRequest;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+/**
+ * Implementers of {@link ViewsModelDecorator} decorate response model with additional entries.
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply views/src/main/java/io/micronaut/views/model/ViewsModelDecorator.java]: @graemerocher changed the java docs

**Comment by sdelamo** [comment]
I've handle the constructor deprecation as pointed by @jameskleeh 

@graemerocher I extracted the method responsible of processing the model to ease `ViewsFilter` override.

**Pull request merged**